lnst/Recipes/ENRT/BaseTunnelRecipe.py

```diff
@@ -35,17 +47,12 @@ def test_wide_configuration(self):
         then the tunnel is created between the specified endpoints in
         :meth:`create_tunnel`.
         """
+        config = self.configure_underlying_network()
+        self.create_tunnel(config)
 
-        configuration = super().test_wide_configuration()
```

Comment:
Right, I get what you mean, but... tunnel recipe config is split into two methods, setup of network devices in `configure_underlying_network()` and the tunnel creation on top of these devices in `create_tunnel()`. If you look at this from a typing perspective, how do we know that the `tunnel_endpoints` property/attribute of the configuration is set when we call `create_tunnel`?

Previously this was not checked. I now changed it so that `configure_underlying_network()` actually creates the configuration object by supplying it with the configured endpoints so that we're sure we have them. If the configuration was created beforehand we wouldn't be able to ensure that we actually properly set these (and we'd be getting a ton of typing errors. I know there's a ton already, but I'm trying to think futureproof).

I guess what we could do is remove `TunnelConfiguration` and make `configure_underlying_network()` return a tuple of configured endpoints make `create_tunnel()` return a tuple of configured tunnel devices. Both of these tuples would then be set in `BaseTunnelRecipe.test_wide_configuration` to the `EnrtConfiguration` so that we don't break the inheritance. Sounds good?

Reply:
Sadly this does not work. The line
```py
config: TunnelConfiguration = super().test_wide_configuration()
```
will display an error because the `EnrtConfiguration` object returned from `BaseEnrtRecipe.test_wide_configuration()` will not conform to the `TunnelConfiguration` protocol. Also I tested this locally and even after setting the tunnel properties to the config object, the object still doesn't conform to the `TunnelConfiguration` protocol. You can try this yourself:
```py
from typing import Protocol


class TunnelConfig(Protocol):
    @property
    def tunnel_endpoints(self) -> tuple:
        ...


class EnrtConfig:
    pass


tunnel_config: TunnelConfig

enrt_config = EnrtConfig()
tunnel_config = enrt_config  # incompatible types error

enrt_config.tunnel_endpoints = (1, 2)  # no attribute error
tunnel_config = enrt_config  # incompatible types error
```
On top of that, this doesn't solve the issue of `create_tunnel` not knowing whether `configure_underlying_network()` actually defined the endpoints in the config.

I'm fairly strongly in favor of the following solution:
```py
class BaseTunnelRecipe(...):
    def test_wide_configuration(self) -> EnrtConfiguration:
        config = super().test_wide_configuration()
        config.tunnel_endpoints = self.configure_underlying_network(config)
        config.tunnel_devices = self.create_tunnel(config.tunnel_endpoints)
        self.wait_tentative_ips(config.tunnel_devices)
        return config

    def configure_underlying_network(self, config: EnrtConfiguration) -> tuple[RemoteDevice, RemoteDevice]:
        ...
        
    def create_tunnel(self, tunnel_endpoints: tuple[RemoteDevice, RemoteDevice]) -> tuple[RemoteDevice, RemoteDevice]:
        ...
```
Note that for some reason `create_tunnel()` adds IPs to tunnel devices, but we never tracked those AFAICS.

Reply:
> Sadly this does not work.

ok, was just a guess... good to know. I guess it makes sense, typing (including this Protocol typing intended for ducktyping) is intended for STATIC analysis... and what my "proposal" would require would be a DYNAMIC type check analysis as the duck typing simply changes over the course of execution.

---

i agree with your final proposal, i think that looks good.

---

if we ever get back to this again i think the proper solution to introducing "recipe specific config class" would have to look like this:

```python
class EnrtConfig(): ...

class BaseEnrtRecipe():
    config_class = EnrtConfig

    @contextmanager
      def _test_wide_context(self):
          config = self.config_class
          self.test_wide_configuration(config)
          self.describe_test_wide_configuration(config)
          try:
              yield config
          finally:
              self.test_wide_deconfiguration(config)


---

class TunnerConfig(EnrtConfig): ...

class TunnelRecipe(...):
    config_class = TunnelConfig
```

this should work and at this point it should be possible to do a correct typing static analysis as well.

it will however require us to update all of our `test_wide_configuration` methods to include `config` as a parameter... but i think that may be a good thing if we want to introduce stricter typing into this.

Reply:
All tunnel recipes have been edited to follow the structure I proposed.

Reply:
One question - why you had to change and needed to explicitly assign the result of these two methods:
```
        config.tunnel_endpoints = self.configure_underlying_network(config)
        config.tunnel_devices = self.create_tunnel(config, config.tunnel_endpoints)
```

Why this assignment can't be done directly in the methods?

Reply:
@jtluka I could, but then we wouldn't be able to enforce that something was actually assigned. Also, it's best to keep functions pure. This function's purpose is to accept a pair of endpoints and create a tunnel on top of them. For me it's a *creation* function and thus should return the created result. Consider the following the functions:
```py
def add(a, b):
    return a + b

def add(obj, a, b):  # replace obj with self, same thing
    obj.attr = a + b
```
I'm for option 1, every time.

Reply:
Ok, let's keep it as is.